### ISSUE TITLE
NullReference Exception in Apply call of a custom Keras model

### ISSUE THREAD

**Issue by liquidiert** [SciSharp/TensorFlow.NET#772]
Hey Guys :D First of all, I really love this project; keep up the good work!

I unfortunately stumbled upon an issue and I'm not quite sure if it is due to me or from the lib.

I've created a custom Keras model (using tf operations and not layers, cause they throw the same error; class vars omitted) :
```c#
public class DQNModel : Model {
  public DQNModel(DQNArgs args) : base(args) {
    var randomNormal = tf.initializers.random_normal_initializer();

    weightsHidden = tf.Variable(randomNormal.Apply(new InitializerArgs((10, 357))));
    biasHidden = tf.Variable(tf.zeros(357));

    weightsOutput = tf.Variable(randomNormal.Apply(new InitializerArgs((357, 4))));
    biasOutput = tf.Variable(tf.zeros(4));

  }

  protected override Tensors Call(Tensors inputs, Tensor state = null, bool is_training = false) {
      try {
          var hiddenLayer = tf.add(tf.matmul(inputs, weightsHidden.AsTensor()), biasHidden.AsTensor());
          hiddenLayer = tf.nn.relu(hiddenLayer);

          var outputLayer = tf.matmul(hiddenLayer, weightsOutput.AsTensor()) + biasOutput.AsTensor();

          return tf.nn.softmax(outputLayer);
      } catch (TensorflowException ex) {
          Console.WriteLine(ex.Message);
          Console.WriteLine(ex.StackTrace);
          Console.WriteLine("-------------------------------------");
          Console.WriteLine(ex.ToString());
          return null;
      }
  }
}
```

Trying something like this:
```c#
tf.enable_eager_execution();
var qVals = net.Apply(input.Tensor);
var maxValIndx = tf.argmax(qVals).numpy(); 
```

Throws then a NullReference error in the [Apply function](https://github.com/SciSharp/TensorFlow.NET/blob/master/src/TensorFlowNET.Keras/Engine/Layer.Apply.cs) of the model layer.

Does anyone have an idea what could be going wrong here? :)

**Comment by Oceania2018** [comment]
@liquidiert Have you taken a look at the [examples](https://github.com/SciSharp/SciSharp-Stack-Examples) before you starting your own model?

**Comment by liquidiert** [comment]
Meanwhile I've managed to get around that error by using an approach similar to the [FCNEager example](https://github.com/SciSharp/SciSharp-Stack-Examples/blob/34bad859574f33fcd3c442ba486051371041c601/src/TensorFlowNET.Examples/NeuralNetworks/FullyConnectedEager.cs).

But now gradient updates via gradient tape won't work :(

```c#
void runOptimization() {
    tf.enable_eager_execution();
    using var g = tf.GradientTape();

    // accumulate data from replay buffer
    // ...

    var forward = net.Apply(states);
    var expand = tf.expand_dims(actions, -1);
    expand = tf.cast(expand, tf.int32); // cast dtype for indices
    var gather = forward.GatherIn1Dim(expand); // an implementation of tf.gather that only uses the first dimension
    var stateActionValues = tf.squeeze(gather);
    var nextStateValues = tf.reduce_max(targetNet.Apply(nextStates));

    var expectedStateActionValues = nextStateValues * GAMMA + rewards;

    var loss = tf.reduce_sum(tf.pow(expectedStateActionValues - stateActionValues, 2f) / (2f * BATCH_SIZE));

    var gradients = g.gradient(loss, net.trainable_variables); // tensors of gradients are null

    optimizer.apply_gradients(zip(gradients, net.trainable_variables.Select(x => x as ResourceVariable)));
}
```

Preceding code works flawless till the gradient update, where it also fails with a NullReference exception:
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Object.GetType()
   at Tensorflow.constant_op.convert_to_eager_tensor(Object value, Context ctx, TF_DataType dtype)
   at Tensorflow.constant_op._constant_impl(Object value, TF_DataType dtype, TensorShape shape, String name, Boolean verify_shape, Boolean allow_broadcast)
   at Tensorflow.ops.convert_to_tensor(Object value, TF_DataType dtype, String name, Boolean as_ref, TF_DataType preferred_dtype, Context ctx)
   at Tensorflow.Eager.EagerRunner.AddInputToOp(Object inputs, Boolean add_type_attr, ArgDef input_arg, List`1 flattened_attrs, List`1 flattened_inputs, SafeOpHandle op, Status status)
   at Tensorflow.Eager.EagerRunner.TFE_FastPathExecute(Context ctx, String device_name, String opName, String name, Action callbacks, Object[] args)
   at Tensorflow.gen_training_ops.resource_apply_adam(Tensor var, Tensor m, Tensor v, Tensor beta1_power, Tensor beta2_power, Tensor lr, Tensor beta1, Tensor beta2, Tensor epsilon, Tensor grad, Boolean use_locking, Boolean use_nesterov, String name)
   at Tensorflow.Keras.Optimizers.Adam._resource_apply_dense(IVariableV1 var, Tensor grad, Dictionary`2 apply_state)
   at Tensorflow.Keras.Optimizers.OptimizerV2.apply_grad_to_update_var(ResourceVariable var, Tensor grad, Dictionary`2 apply_state)
   at Tensorflow.Keras.Optimizers.OptimizerV2.<>c__DisplayClass23_1.<_distributed_apply>b__1(NameScope <p0>)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.Keras.Optimizers.OptimizerV2.<>c__DisplayClass23_0.<_distributed_apply>b__0(NameScope <p0>)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.Keras.Optimizers.OptimizerV2._distributed_apply(IEnumerable`1 grads_and_vars, String name, Dictionary`2 _apply_state)
   at Tensorflow.Keras.Optimizers.OptimizerV2.<>c__DisplayClass20_0.<apply_gradients>b__1(NameScope <p0>)
   at Tensorflow.Binding.tf_with[TIn,TOut](TIn py, Func`2 action)
   at Tensorflow.Keras.Optimizers.OptimizerV2.apply_gradients(IEnumerable`1 grads_and_vars, String name, Boolean experimental_aggregate_gradients)
   at lean_pactheman_client.DQN.DQN.runOptimization() in /home/secretone/git_repos/uni/lean-pactheman-client/cSharp/Interfaces/DQN/DQN.cs:line 132
   at lean_pactheman_client.DQN.DQN.PerformMoveAsync(PlayerInfo info) in /home/secretone/git_repos/uni/lean-pactheman-client/cSharp/Interfaces/DQN/DQN.cs:line 160
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

@Oceania2018 any ideas why the gradient calculation fails? I also took a look at the GradientTape object and made sure it's recording (which it is btw). I'm kinda clueless on that one to :/

**Comment by Oceania2018** [comment]
@liquidiert Please try latest version.
